src/core/struct_tree.js

```diff
@@ -67,6 +69,425 @@ class StructTreeRoot {
       this.roleMap.set(key, value.name);
     });
   }
+
+  static async canCreateStructureTree({
+    catalogRef,
+    pdfManager,
+    newAnnotationsByPage,
+  }) {
+    if (!(catalogRef instanceof Ref)) {
+      warn("Cannot save the struct tree: no catalog reference.");
+      return false;
+    }
+
+    const cleanup = map => {
+      for (const elements of map.values()) {
+        for (const element of elements) {
+          delete element.parentTreeId;
+        }
+      }
+    };
+
+    let nextKey = 0;
+    let hasNothingToUpdate = true;
+
+    for (const [pageIndex, elements] of newAnnotationsByPage) {
+      const { ref: pageRef } = await pdfManager.getPage(pageIndex);
+      if (!(pageRef instanceof Ref)) {
+        cleanup(newAnnotationsByPage);
+        warn(`Cannot save the struct tree: page ${pageIndex} has no ref.`);
+        return false;
```

Comment:
We could perhaps "inline" the cleanup-helper if we instead do the following here:
```suggestion
        warn(`Cannot save the struct tree: page ${pageIndex} has no ref.`);
        hasNothingToUpdate = true;
        break;
```